src/pages/workspaces/workspace/WorkspaceMenu.ts

```diff
@@ -0,0 +1,174 @@
+import { icon } from '@terra-ui-packages/components';
+import { cond, DEFAULT } from '@terra-ui-packages/core-utils';
+import { Fragment } from 'react';
+import { h } from 'react-hyperscript-helpers';
+import { Clickable } from 'src/components/common';
+import { MenuButton } from 'src/components/MenuButton';
+import { makeMenuIcon, MenuTrigger } from 'src/components/PopupTrigger';
+import { useWorkspaceDetails } from 'src/components/workspace-utils';
+import { isOwner, WorkspacePolicy, WorkspaceWrapper as Workspace } from 'src/libs/workspace-utils';
+
+const isNameType = (o: WorkspaceInfo): o is DynamicWorkspaceInfo =>
+  'name' in o && typeof o.name === 'string' && 'namespace' in o && typeof o.namespace === 'string';
+
+type LoadedWorkspaceInfo = { canShare: boolean; isLocked: boolean; isOwner: boolean; workspaceLoaded: boolean };
+type DynamicWorkspaceInfo = { name: string; namespace: string };
+type WorkspaceInfo = DynamicWorkspaceInfo | LoadedWorkspaceInfo;
+
+interface WorkspaceMenuCallbacks {
+  onClone: () => void;
+  onShare: (policies?: WorkspacePolicy[]) => void;
+  onLock: () => void;
+  onDelete: () => void;
+  onLeave: () => void;
+}
+
+interface WorkspaceMenuProps {
+  iconSize?: number;
+  popupLocation: unknown; // will error whenever PopupTrigger.js is converted to ts
+  callbacks: WorkspaceMenuCallbacks;
+  workspaceInfo: WorkspaceInfo;
+}
+// In `workspaceInfo`, specify either `name and namespace` to fetch the Workspace details,
+// or `canShare, isLocked, and isOwner` to use previously fetched details.
+export const WorkspaceMenu = (props: WorkspaceMenuProps) => {
```

Comment:
Nit: preferably, include the return type for components.

```suggestion
export const WorkspaceMenu = (props: WorkspaceMenuProps): ReactNode => {
```